File: src/transport/poller.ts
// Telegram long-poll transport. Replaces the inbound webhook server: the box makes OUTBOUND
// getUpdates calls and never listens on a port, so the bot runs behind NAT, on a home box, or on a
// bare VM with no public URL or TLS. One consumer loop pulls updates, hands each to `onUpdate`
// (which authorizes + enqueues), and advances the confirmation offset past it.

import type { GetUpdatesOptions, TelegramUpdate } from "./telegram.js";
import { logger } from "../logger.js";

export interface Poller {
  /** Stop after the in-flight long-poll returns (the abort makes that immediate). */
  stop(): Promise<void>;
}

export interface PollerDeps {
  getUpdates: (opts?: GetUpdatesOptions) => Promise<TelegramUpdate[]>;
  /** Called for each update in order. Must be fast and non-throwing (it enqueues and returns). */
  onUpdate: (update: TelegramUpdate) => void;
  /** Long-poll timeout in seconds (Telegram holds the request open this long when idle). */
  timeoutSeconds?: number;
  /** Backoff after a failed poll, in ms (network errors); default 1000. */
  backoffMs?: number;
}

export function startPoller(deps: PollerDeps): Poller {
  const timeout = deps.timeoutSeconds ?? 50;
  const backoffMs = deps.backoffMs ?? 1000;
  const controller = new AbortController();
  let stopped = false;
  let offset: number | undefined;

  async function loop(): Promise<void> {
    while (!stopped) {
      try {
        const updates = await deps.getUpdates({
          ...(offset !== undefined ? { offset } : {}),
          timeout,
          signal: controller.signal,
        });
        for (const update of updates) {
          if (update.update_id !== undefined) offset = update.update_id + 1;
          try {
            deps.onUpdate(update);
          } catch (err) {
            logger.error("onUpdate failed", { error: (err as Error).message });
          }
        }
      } catch (err) {
        if (stopped) break; // the abort during shutdown surfaces here — expected, not an error.
        logger.error("getUpdates failed; backing off", { error: (err as Error).message });
        await sleep(backoffMs);
      }
    }
  }

  const done = loop();
  logger.info("Telegram long-poll started", { timeout });

  return {
    async stop(): Promise<void> {
      stopped = true;
      controller.abort();
      await done.catch(() => undefined);
    },
  };
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: src/transport/poller.rs
//! Telegram long-poll transport. Port of `src/transport/poller.ts`. The box makes OUTBOUND
//! getUpdates calls and never listens on a port. One loop pulls updates, forwards each over a
//! channel (the app authorizes + enqueues), and advances the confirmation offset past it.

use std::sync::Arc;
use std::time::Duration;

use tokio::sync::Notify;
use tokio::sync::mpsc::UnboundedSender;

use super::telegram::{TelegramClient, TelegramUpdate};

/// Default long-poll timeout (Telegram holds the request open this long when idle).
const DEFAULT_TIMEOUT_SECS: u32 = 50;
/// Backoff after a failed poll.
const BACKOFF: Duration = Duration::from_millis(1000);

/// Run the poll loop until `shutdown` is notified. Each update is forwarded to `tx`.
pub async fn run(
    client: TelegramClient,
    tx: UnboundedSender<TelegramUpdate>,
    shutdown: Arc<Notify>,
) {
    let mut offset: Option<i64> = None;
    tracing::info!(timeout = DEFAULT_TIMEOUT_SECS, "Telegram long-poll started");
    loop {
        tokio::select! {
            _ = shutdown.notified() => break,
            res = client.get_updates(offset, DEFAULT_TIMEOUT_SECS) => match res {
                Ok(updates) => {
                    for update in updates {
                        if let Some(id) = update.update_id {
                            offset = Some(id + 1);
                        }
                        if tx.send(update).is_err() {
                            return; // app loop gone — stop polling
                        }
                    }
                }
                Err(err) => {
                    tracing::error!(%err, "getUpdates failed; backing off");
                    tokio::select! {
                        _ = shutdown.notified() => break,
                        _ = tokio::time::sleep(BACKOFF) => {}
                    }
                }
            }
        }
    }
    tracing::info!("Telegram long-poll stopped");
}
